File: CodeTopGo/46.permute.go
/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 */

// @lc code=start
var allPossible [][]int
func permute(nums []int) [][]int{
	allPossible = [][]int{}
	backPermute(nums, 0)
	return allPossible
}

func backPermute(nums []int, first int){
	if first == len(nums){
		curr := make([]int, len(nums))
		copy(curr, nums)
		allPossible = append(allPossible, curr)
		// allPossible = append(allPossible, nums) 之前是这样写，但是最后的 数组都是一样的， 原因为直接这样赋值 allPossible 存储的是nums的地址，当nums变化时 allPossible中的数组也跟着变化。
	}
	for i := first; i < len(nums); i++{
		swap(nums, first, i)
		backPermute(nums, first + 1)
		swap(nums, first, i)
	}
}
func swap(nums []int, a int, b int){
	nums[a], nums[b] = nums[b], nums[a]
}

// var res [][]int
// func permute(nums []int) [][]int {
// 	res = [][]int{}
// 	backTrack(nums,len(nums),[]int{})
// 	return res
// }
// func backTrack(nums []int,numsLen int,path []int)  {
// 	if len(nums)==0{
// 		p:=make([]int,len(path))
// 		copy(p,path)
// 		res = append(res,p)
// 	}
// 	for i:=0;i<numsLen;i++{
// 		cur:=nums[i]
// 		path = append(path,cur)
// 		nums = append(nums[:i],nums[i+1:]...)//直接使用切片
// 		backTrack(nums,len(nums),path)
// 		nums = append(nums[:i],append([]int{cur},nums[i:]...)...)//回溯的时候切片也要复原，元素位置不能变
// 		path = path[:len(path)-1]

// 	}
// }

// @lc code=end

File: CodeTop/46.permute.cpp
/*
 * @lc app=leetcode.cn id=46 lang=cpp
 *
 * [46] 全排列
 * 给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。
 * [medium]
 * https://leetcode-cn.com/problems/permutations/
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> numsList;
        backtrack(nums, numsList, 0);
        return numsList;
    }
    void backtrack(vector<int>& nums, vector<vector<int>>& numsList, int first){
        if(first == nums.size()){
            numsList.push_back(nums);
        }
        for(int i = first; i < nums.size(); i++){
            swap(nums[i], nums[first]);
            backtrack(nums, numsList, first + 1);
            swap(nums[i], nums[first]);
        }
    }
};
// @lc code=end

#include <iostream>
#include <vector>

using namespace std;
//背就vans了
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> numsList;
        backtrack(nums,numsList,0);
        return numsList;
    }

    void backtrack(vector<int>& nums,vector<vector<int>>& numsList,int first){
        if(first==nums.size()){
            numsList.push_back(nums);
        }

        for(int i=first;i<nums.size();i++){
            swap(nums[first],nums[i]);
            backtrack(nums,numsList,first+1);
            swap(nums[first],nums[i]);
        }
    }
};